prometheus-net/Histogram.cs

```diff
@@ -106,28 +85,15 @@ public void Observe(double val)
                     return;
                 }
 
-                int bucketIndex = -1;
-                for (int i = 0; i < _buckets.Length; i++)
+                for (int i = 0; i < _upperBounds.Length; i++)
                 {
-                    if (val <= _buckets[i])
+                    if (val <= _upperBounds[i])
                     {
-                        bucketIndex = i;
+                        _bucketCounts[i].Add(1);
                         break;
                     }
                 }
-
-                if (bucketIndex == -1)
-                {
-                    Trace.WriteLine(string.Format("Couldn't find a bucket for {0} in metric {1}", val, ToString()));
-                    return;
-                }
-
-                lock (_lock)
```

Comment:
From Prometheus client library authoring guidance:
_In our experience the least performant is mutexes._
_Processor atomic instructions tend to be in the middle, and generally acceptable._
_Approaches that avoid different CPUs mutating the same bit of RAM work best, such as the DoubleAdder in Java’s simpleclient. There is a memory cost though._

By converting the bucket counts and sum to ThreadSafeDouble we can avoid the need to wrap this in a lock statement.